viz_scripts/scaffolding.py

```diff
@@ -311,29 +356,123 @@ def energy_impact_kWH(df,distance):
+    df['CO2_Impact(lb)'] = round((df['Replaced_mode_lb_CO2'] - df['Mode_confirm_lb_CO2']), 3)
+
+    # Convert the CO2_Impact to be represented in kilogram
+    df['CO2_Impact(kg)']  = round((df['Replaced_mode_kg_CO2'] - df['Mode_confirm_kg_CO2']), 3)
     return df
+
+def compute_CO2_footprint_dynamic(expanded_ct, dynamic_labels):
+    conversion_meter_to_kilometer = 0.001
+    conversion_kilogram_to_lbs = 2.20462
+    conversion_kilometer_to_mile = 0.621371
+
+    dic_mode_kgCO2PerKm = {mode["value"]: mode["kgCo2PerKm"] for mode in dynamic_labels["MODE"]}
+
+    if "mode_confirm" in expanded_ct.columns:
+        expanded_ct['Mode_confirm_kg_CO2'] = ((expanded_ct['distance'] * conversion_meter_to_kilometer )) * (expanded_ct['mode_confirm'].map(dic_mode_kgCO2PerKm))
+        expanded_ct['Mode_confirm_kg_CO2'] = expanded_ct['Mode_confirm_kg_CO2'].fillna(0)
+        expanded_ct['Mode_confirm_lb_CO2'] = expanded_ct['Mode_confirm_kg_CO2'] * conversion_kilogram_to_lbs
+
+    return expanded_ct
+
+def compute_CO2_impact_dynamic(expanded_ct, dynamic_labels):
+    conversion_meter_to_kilometer = 0.001 # 1 meter = 0.001 kilometer
+    conversion_kilogram_to_lbs = 2.20462
+    conversion_kilometer_to_mile = 0.621371
+
+    dic_mode_kgCO2PerKm = {mode["value"]: mode["kgCo2PerKm"] for mode in dynamic_labels["MODE"]}
+
+    if "mode_confirm" in expanded_ct.columns:
+        # The expanded_ct['Mode_confirm_kg_CO2'] is CO2 emission in terms of [distance in kms * CO2 emission in kgCO2 per km = kg of CO2]
+        expanded_ct['Mode_confirm_kg_CO2'] = ((expanded_ct['distance'] * conversion_meter_to_kilometer )) * (expanded_ct['mode_confirm'].map(dic_mode_kgCO2PerKm))
+        expanded_ct['Mode_confirm_kg_CO2'] = expanded_ct['Mode_confirm_kg_CO2'].fillna(0)
+        expanded_ct['Mode_confirm_lb_CO2'] = expanded_ct['Mode_confirm_kg_CO2'] * conversion_kilogram_to_lbs
+
+    if "replaced_mode" in expanded_ct.columns:
+        expanded_ct['Replaced_mode_kg_CO2'] = (expanded_ct['distance'] * conversion_meter_to_kilometer ) * (expanded_ct['replaced_mode'].map(dic_mode_kgCO2PerKm))
+        expanded_ct['Replaced_mode_kg_CO2'] = expanded_ct['Replaced_mode_kg_CO2'].fillna(0)
+        expanded_ct['Replaced_mode_lb_CO2'] = expanded_ct['Replaced_mode_kg_CO2'] * conversion_kilogram_to_lbs
+
+    expanded_ct['CO2_Impact(kg)']  = round ((expanded_ct['Replaced_mode_kg_CO2'] - expanded_ct['Mode_confirm_kg_CO2']), 3)
+    expanded_ct['CO2_Impact(lb)'] = round ((expanded_ct['Replaced_mode_lb_CO2'] - expanded_ct['Mode_confirm_lb_CO2']), 3)
+    return expanded_ct
+
+def print_CO2_emission_calculations(data_eb, ebco2_lb, ebco2_kg, dynamic_labels):
+
+    filtered_taxi_data = data_eb[data_eb['Replaced_mode'] == "Taxi/Uber/Lyft"]
+    filtered_bus_data = data_eb[data_eb['Replaced_mode'] == "Bus"]
+    filtered_freeshuttle_data = data_eb[data_eb['Replaced_mode'] == "Free Shuttle"]
+    filtered_walk_data = data_eb[data_eb['Replaced_mode'] == "Walk"]
```

Comment:
I am not going to require any changes here since this is only invoked in debug cases, but note that dynamic configs are not guaranteed to have the same labels, or even the same modes. Note that the Lao labels, for example that I asked you to test against, do not have the translation "Gas Car Shared Ride".

Reply:
Please add some notes here about the particular dynamic config that this works with. Because we have at least two production (non-example) dynamic configs now (Laos and CA e-bike) and the second is even in English. So people might be confused while debugging and not know why this doesn't work.

Reply:
Updated the comment by mentioning the print statements for debugging mode are specifically designed in consideration with [https://github.com/e-mission/nrel-openpath-deploy-configs/blob/main/configs/dev-emulator-program.nrel-op.json](https://github.com/e-mission/nrel-openpath-deploy-configs/blob/main/configs/dev-emulator-program.nrel-op.json).